kubeadm: add task pages for adding Linux and Windows worker nodes (#47888)
kubernetes/website: content/en/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes.md

---
title: Adding Windows worker nodes
content_type: task
weight: 50
---

<!-- overview -->

This page explains how to add Windows worker nodes to a kubeadm cluster.

## {{% heading "prerequisites" %}}

* A running [Windows Server 2022](https://www.microsoft.com/cloud-platform/windows-server-pricing)
(or higher) instance with administrative access.
* A running kubeadm cluster created by `kubeadm init` and following the steps
in the document [Creating a cluster with kubeadm](/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/).

<!-- steps -->

## Adding Windows worker nodes

{{< note >}}
To facilitate the addition of Windows worker nodes to a cluster, PowerShell scripts from the repository
https://sigs.k8s.io/sig-windows-tools are used.
{{< /note >}}

Do the following for each machine:

1. Connect to the machine by using SSH or another method.
1. Make sure you are Administrator or a privileged user.

> 1. Open a PowerShell session.

Reviewer: (nit) You can [run PowerShell locally](https://learn.microsoft.com/en-us/powershell/scripting/learn/ps101/08-powershell-remoting) and invoke the commands remotely.

Author: reworded to tell the user to "open powershell session on the machine"

Then proceed with the steps outlined below.

### Install containerd

> 

Reviewer: This needs a third party content callout. containerd is not part of Kubernetes.

```

By default, node join tokens expire after 24 hours. If you are joining a node to the cluster after the
current token has expired, you can create a new token by running the following command on the
control plane node:

```bash
kubeadm token create
```

The output is similar to this:

```console
5didvk.d09sbcov8ph2amjw
```

If you don't have the value of `--discovery-token-ca-cert-hash`, you can get it by running the
following commands on the control plane node:

```bash
sudo cat /etc/kubernetes/pki/ca.crt | openssl x509 -pubkey  | openssl rsa -pubin -outform der 2>/dev/null | \
   openssl dgst -sha256 -hex | sed 's/^.* //'
```

The output is similar to:

```console
8cb2de97839780a412b93877f8507ad6c94f73add17d5d7058e91741c9d5ec78
```

The output of the `kubeadm join` command should look something like:

```
[preflight] Running pre-flight checks

... (log output of join workflow) ...

Node join complete:
* Certificate signing request sent to control-plane and response
  received.
* Kubelet informed of new secure connection details.

Run 'kubectl get nodes' on control-plane to see this machine join.
```

> A few seconds later, you should notice this node in the output from `kubectl get nodes`
> when run on the control plane node.

Reviewer: You can run `kubectl` elsewhere

(nit)
```suggestion
A few seconds later, you should notice this node in the output from `kubectl get nodes`.
(for example, run `kubectl` on a  control plane node).
```

### Network configuration

CNI setup on clusters mixed with Linux and Windows nodes requires more steps than just
running `kubectl apply` on a manifest file. Additionally, the CNI plugin running on control
plane nodes must be prepared to support the CNI plugin running on Windows worker nodes.

{{% thirdparty-content %}}

Only a few CNI plugins currently support Windows. Below you can find individual setup instructions for them:
* [Flannel](https://sigs.k8s.io/sig-windows-tools/guides/flannel.md)
* [Calico](https://docs.tigera.io/calico/latest/getting-started/kubernetes/windows-calico/)

> ### Install kubectl for Windows (optional)

Reviewer: (nit)
```suggestion
### Install kubectl for Windows (optional) {#install-kubectl}
```

---
title: Adding Windows worker nodes
content_type: task
weight: 50
---

<!-- overview -->

This page explains how to add Windows worker nodes to a kubeadm cluster.

## {{% heading "prerequisites" %}}

> * A running [Windows Server 2019](https://www.microsoft.com/en-us/cloud-platform/windows-server-pricing)

Reviewer 2: We should probably point to WS2022, as 2019 is in extended support

Author: should i replace the text in squire brackets to say `[WIndows Server 2022]` instead of 2019?

> * A running [Windows Server 2019](https://www.microsoft.com/en-us/cloud-platform/windows-server-pricing)

Reviewer 3: ```suggestion
* A running [Windows Server 2019](https://www.microsoft.com/cloud-platform/windows-server-pricing)
```

(or higher) instance with administrative access.
* If you are using VXLAN/Overlay networking you must have also have
[KB4489899](https://support.microsoft.com/help/4489899) installed.
* A running kubeadm cluster created by `kubeadm init` and following the steps
in the document [Creating a cluster with kubeadm](/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/).

> * `curl.exe` is installed and present in the `PATH` environment variable.

Reviewer 3: I think we can drop this point too. Curl.exe has been included in Windows since Windows Server 2019 which is the minimum target we have for Windows nodes.

<!-- steps -->

## Adding Windows worker nodes

{{< note >}}
To facilitate the addition of Windows worker nodes to a cluster, PowerShell scripts from the repository
https://sigs.k8s.io/sig-windows-tools are used.
{{< /note >}}

Do the following for each machine:

1. Connect to the machine by using SSH or another method.
1. Make sure you are Administrator or a privileged user.
1. Open a PowerShell session.

Then proceed with the steps outlined below.

### Install containerd

Run the following commands to install containerd:

  ```PowerShell
  curl.exe -LO https://raw.githubusercontent.com/kubernetes-sigs/sig-windows-tools/master/hostprocess/Install-Containerd.ps1

>   .\Install-Containerd.ps1 -ContainerDVersion 1.7.1

Reviewer 2: ```suggestion
  .\Install-Containerd.ps1 -ContainerDVersion  1.7.22
```

Reviewer 2: should we point to the containerd runtime docs here instead? https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd

Author: i did not notice that there are windows install steps in https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd (inside the getting started link) that's great because we can just add a prereq for a container runtime.

ideally we should break the follwing into linux/windows eventually too:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl

and maybe have PrepareNode.ps be the install step for windows in that guide.

Reviewer 3: the biggest differences between those docs are what we have are

- Checking and enabling the Hyper-V feature - which I don't think we care about for the purposes of this document 
- Setting the CNI bin and CNI config dir's in the containerd config - I think we do want this

I think if we can get the containerd docs updated to create and set the cni bin/config dirs in the containerd setup script we can move away from the containerd setup doc in sig-windows-tools.
@jsturtevant - WDYT?

Reviewer 2: agree, That info should live in the containerd install doc

Author: i though about unrolling the install-containerd script but think i'm going to leave this task to SIG Windows. my main problem here is that i have no way to test the containerd install and CNI setup, ATM.

also, it seems out of scope of the PR.

>   .\Install-Containerd.ps1 -ContainerDVersion 1.7.1

Reviewer 3: we should probably not recommend a specific version - i've seen this guidance in other areas of the K8s docs

Author: added 1.7.22 since IIUC there has to be a version here or it will use an older version as pinned inside the script. in this case 1.7.1, but we should probably decide what to do with this thread https://github.com/kubernetes/website/pull/47888#discussion_r1755223274

Reviewer: It's OK to recommend a version provided we have the capacity to update this every time the other project does a release (we probably don't).

Author: reworked the steps to tell the user to pick a release version from the containerd repo.

```

The output is similar to:

```console
8cb2de97839780a412b93877f8507ad6c94f73add17d5d7058e91741c9d5ec78
```

The output of the `kubeadm join` command should look something like:

```
[preflight] Running pre-flight checks

... (log output of join workflow) ...

Node join complete:
* Certificate signing request sent to control-plane and response
  received.
* Kubelet informed of new secure connection details.

Run 'kubectl get nodes' on control-plane to see this machine join.
```

A few seconds later, you should notice this node in the output from `kubectl get nodes`
when run on the control plane node.

### Network configuration

CNI setup on clusters mixed with Linux and Windows nodes requires more steps than just
running `kubectl apply` on a manifest file. Additionally, the CNI plugin running on control
plane nodes must be prepared to support the CNI plugin running on Windows worker nodes.

{{% thirdparty-content %}}

Only a few CNI plugins currently support Windows. Below you can find individual setup instructions for them:
* [Flannel](https://sigs.k8s.io/sig-windows-tools/guides/flannel.md)
* [Calico](https://sigs.k8s.io/sig-windows-tools/guides/calico.md)

> * [Calico Operator](https://sigs.k8s.io/sig-windows-tools/guides/calico_operator.md)

Reviewer 2: I'd prefer to link to calico's official installation scripts and not including links to the sig-windows guides.  They have gone stale and community hasn't been able to maintain them

https://docs.tigera.io/calico/latest/getting-started/kubernetes/windows-calico/

---

> title: Adding Linux worker nodes

Reviewer: Did you mean Windows?

content_type: task
weight: 50
---

<!-- overview -->

> This page explains how to add Linux worker nodes to a kubeadm cluster.

Reviewer: Did you mean Windows?

## {{% heading "prerequisites" %}}

* Obtain a [Windows Server 2019 license](https://www.microsoft.com/en-us/cloud-platform/windows-server-pricing) (or higher) in order to configure the Windows node that hosts Windows containers. If you are using VXLAN/Overlay networking you must have also have [KB4489899](https://support.microsoft.com/help/4489899) installed.
* `curl.exe` is installed and present in the `PATH` environment variable.
* A running kubeadm cluster created by `kubeadm init` and following the steps

> in the document [Creating a cluster with kubeadm](/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/).

Author: not clear what the problem is?

"creating a cluster" is the landing page where users must call "init" to create the control plane on Linux.
it's a prereq for Windows nodes to be added.

> in the document [Creating a cluster with kubeadm](/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/).

Reviewer 3: yes, we currently don't support running the control plane components on Windows (and don't have any plans to do so in the future).

> 

Reviewer: You'll also need a server running Windows, and administrative access to the server (just having the license is not enough to proceed).

Author: attempted to reword the steps.

<!-- steps -->

## Adding Windows worker nodes

{{< note >}}
To facilitate the addition of Windows worker nodes to a cluster, PowerShell scripts from the repository
https://sigs.k8s.io/sig-windows-tools are used. A long term goal is to simplify the addition of Windows
worker nodes and align its complexity to
[Adding Linux worker nodes](/docs/tasks/administer-cluster/kubeadm/adding-linux-nodes/).
{{< /note >}}

Do the following for each machine:

1. Connect to the machine by using SSH or another method.
1. Make sure you are Administrator or a privileged user.
1. Open a PowerShell session.

Then proceed with the steps outlined below.

### Install containerd

Run the following commands to install containerd:

  ```PowerShell

>   curl.exe -LO https://raw.githubusercontent.com/kubernetes-sigs/sig-windows-tools/master/hostprocess/Install-Containerd.ps1

Author: i think yes, that's how it was originally.

>   curl.exe -LO https://raw.githubusercontent.com/kubernetes-sigs/sig-windows-tools/master/hostprocess/Install-Containerd.ps1

Reviewer 3: `master` is OK with me.
We'll make sure that script works across all supported version of Windows

  ```

#### Additional information about kubeadm join

{{< note >}}
To specify an IPv6 tuple for `<control-plane-host>:<control-plane-port>`, IPv6 address must be enclosed in square brackets, for example: `[2001:db8::101]:2073`.
{{< /note >}}

If you do not have the token, you can get it by running the following command on the control plane node:

```bash
kubeadm token list
```

The output is similar to this:

```console
TOKEN                    TTL  EXPIRES              USAGES           DESCRIPTION            EXTRA GROUPS
8ewj1p.9r9hcjoqgajrj4gi  23h  2018-06-12T02:51:28Z authentication,  The default bootstrap  system:
                                                   signing          token generated by     bootstrappers:
                                                                    'kubeadm init'.        kubeadm:
                                                                                           default-node-token
```

By default, tokens expire after 24 hours. If you are joining a node to the cluster after the current token has expired. You can create a new token by running the following command on the control plane node:

```bash

> kubeadm token create

Author: updated, minus `# Run this on a control plane node`. the prev sentence already mentions it.

> kubeadm token create

Reviewer 3: Should we move the steps about creating kubeadm join tokens to `install-kubeadm.md` and reference that from both the windows and linux docs?
It looks like some of the docs are duplicated here

```

The output is similar to this:

```console
5didvk.d09sbcov8ph2amjw
```

If you don't have the value of `--discovery-token-ca-cert-hash`, you can get it by running the
following commands on the control plane node:

```bash
openssl x509 -pubkey -in /etc/kubernetes/pki/ca.crt | openssl rsa -pubin -outform der 2>/dev/null | \
   openssl dgst -sha256 -hex | sed 's/^.* //'
```

The output is similar to:

```console
8cb2de97839780a412b93877f8507ad6c94f73add17d5d7058e91741c9d5ec78
```

The output of the `kubeadm join` command should look something like:

```
[preflight] Running pre-flight checks

... (log output of join workflow) ...

Node join complete:
* Certificate signing request sent to control-plane and response
  received.
* Kubelet informed of new secure connection details.

Run 'kubectl get nodes' on control-plane to see this machine join.
```

A few seconds later, you should notice this node in the output from `kubectl get nodes`
when run on the control plane node.

Install kubectl for windows (optional)

> For more information about it : https://kubernetes.io/docs/tasks/tools/install-kubectl-windows/

Author: deleted this as it was a duplicate.
already present in a new `### Install kubectl for Windows (optional)` near the bottom.

### Network configuration

CNI setup on clusters mixed with Linux and Windows nodes requires more steps than just
running `kubectl apply` on a manifest file. Additionally, the CNI plugin running on control
plane nodes must be prepared to support the CNI plugin running on Windows worker nodes.

> This is another point of improvement for the future.

Reviewer: ```suggestion
```

We avoid statements about the future.
